perma_web/fabfile/dev.py

```diff
@@ -573,3 +573,113 @@ def check_s3_hashes():
                 m.update(buf)
         if m.hexdigest() != remote_paths[local_path]:
             print "Hash mismatch! Local: %s Remote: %s" % (m.hexdigest(), remote_paths[local_path])
+
+
+@task
+def check_storage():
+    """
+        Confirm that, for every link, there is a WARC in each storage, and that their hashes match.
+
+        Ground truth is the list of link objects: compare its list of warcs with those of each storage,
+        and compare hashes when more than one such file is present.
+
+        Derived from check_s3_hashes
+    """
+    from django.core.files.storage import default_storage
+    from django.db.models import Q
+    from perma.models import Link, Capture
+    from tqdm import tqdm
+
+    # this can be generalized later to an arbitrary number of storages
+    storages = {'primary': {'storage': default_storage, 'lookup': {}}}
+    if hasattr(default_storage, 'secondary_storage'):
+        storages.update({'secondary': {'storage': default_storage.secondary_storage, 'lookup': {}}})
+
+    # only use cache files when all are present: link cache, and one for each storage
+    link_cache = '/tmp/perma_link_cache.txt'
+    caches = [link_cache]
+    for key in storages:
+        caches.append('/tmp/perma_storage_cache_{0}.txt'.format(key))
+
+    if not all(map(os.path.exists, caches)):
+        print("Building link cache ...")
+        with open(link_cache, 'w') as tmp_file:
+            capture_filter = (Q(role="primary") & Q(status="success")) | (Q(role="screenshot") & Q(status="success"))
+            for link in tqdm(Link.objects.filter(captures__in=Capture.objects.filter(capture_filter)).distinct()):
+                tmp_file.write("{0}\n".format(link.warc_storage_file()))
+                # this produces strings like u'warcs/0G/GO/XR/XG/0-GGOX-RXGQ.warc.gz'; make the storage paths match
+                # by chopping off the prefix, whether storage.location, ._root_path, or .base_location
+        print("Building storage cache{0} ...".format("s" if len(storages) > 1 else ""))
+        for key in storages:
+            storage = storages[key]['storage']
+            with open('/tmp/perma_storage_cache_{0}.txt'.format(key), 'w') as tmp_file:
+                if hasattr(storage, 'bucket'):
+                    # S3
+                    for f in tqdm(storage.bucket.list('generated/warcs/')):
+                        # etag is a string like u'"3ea8c903d9991d466ec437d1789379a6"', so we need to
+                        # knock off the extra quotation marks
+                        # note that etags are not always md5sums, but should be in these cases; we can rewrite
+                        # or replace md5hash if necessary
+                        path = f.key[(len(storage.location)):]
```

Comment:
Is this is where we are "chopping off the prefix, whether storage.location..."? Total nitpick: I personally would find it easier to read this code if that comment were here, modifying this line. It's not that the concept is confusing; I think I was thrown because the comment and the code are in such different scopes. No need to address if you disagree.